templates/simplerule.epp

```diff
@@ -0,0 +1,55 @@
+<%- | String                  $action,
+      Optional[String]        $comment,
+      Optional[Variant[Array[Stdlib::Port, 1], Stdlib::Port, String]] $dport,
+      Optional[String]        $proto,
+      Optional[Variant[Stdlib::IP::Address::V6, Stdlib::IP::Address::V4, Pattern[/^@[-a-zA-Z0-9_]+$/]]] $daddr,
+      Enum['ip', 'ip6']       $set_type,
+      Boolean                 $counter,
+| -%>
+<%- if $proto {
+  $_proto = $proto ? {
+    /tcp(4|6)?/ => 'tcp',
+    /udp(4|6)?/ => 'udp',
+  }
+  $_ip_version_filter = $proto ? {
+    /(tcp4|udp4)/ => 'ip version 4',
+    /(tcp6|udp6)/ => 'ip version 6',
+    default       => undef,
+  }
+} else {
+  $_ip_version_filter = undef
+} -%>
+<%- if $daddr {
+  if $daddr =~ Stdlib::IP::Address::V6 {
+    $_dst_hosts = "ip6 daddr ${daddr}"
+  } elsif $daddr =~ Stdlib::IP::Address::V4 {
+    $_dst_hosts = "ip daddr ${daddr}"
+  } else {
+    $_dst_hosts = $set_type ? {
+      'ip'  => "ip daddr ${daddr}",
+      'ip6' => "ip6 daddr ${daddr}",
+    }
+  }
+} else {
+  $_dst_hosts = undef
+} -%>
+<%- if $proto and $dport {
+  if $dport =~ Array {
+    $_dst_port = "${_proto} dport {${dport.join(', ')}}"
+  } else {
+    $_dst_port = "${_proto} dport $dport"
+  }
```

Comment:
Just a suggestion for the future: I've seen this pattern in a few places. Perhaps a separate function should be introduced to format an array/string as nft accepts it.

Reply:
or we could make sure dport is always an array meaning to wrap it as an array and then always use `{${dport.join(', ')}}`

But pay attention since dport can be integers you can't do a simple `Array($dport)` as this has a different semantic than `any2array($dport)` from stdlib.

so `$_dst_port = "${_proto} dport {${any2array($dport).join(', ')}}"`

Reply:
Yes, we could keep `Optional[Variant[Array[Stdlib::Port, 1], Stdlib::Port, Pattern[/\d+-\d+/]]]` as type and treat it always as an array.

> But pay attention since dport can be integers you can't do a simple Array($dport) as this has a different semantic than any2array($dport) from stdlib.

This should do the job using `Array()`, note the `true` ([docs](https://github.com/puppetlabs/puppetlabs-stdlib/blob/main/REFERENCE.md#any2array-1)):

```
$_dst_port = "${_proto} dport {${Array($dport, true).join(', ')}}"
```

In action:

```puppet
notify{"tcp dport {${Array(53, true).join(', ')}}":}
notify{"tcp dport {${Array([53, 54], true).join(', ')}}":}
notify{"tcp dport {${Array('53-54', true).join(', ')}}":}
```

Output:

```
Notice: Compiled catalog for example in environment production in 0.03 seconds
Notice: tcp dport {53}
Notice: /Stage[main]/Main/Notify[tcp dport {53}]/message: defined 'message' as 'tcp dport {53}'
Notice: tcp dport {53, 54}
Notice: /Stage[main]/Main/Notify[tcp dport {53, 54}]/message: defined 'message' as 'tcp dport {53, 54}'
Notice: tcp dport {53-54}
Notice: /Stage[main]/Main/Notify[tcp dport {53-54}]/message: defined 'message' as 'tcp dport {53-54}'
Notice: Applied catalog in 0.13 seconds
```

The three above are valid syntax-wise for nftables.